Lib/test/test_asyncgen.py

```diff
@@ -1798,6 +1798,12 @@ async def run():
         res = self.loop.run_until_complete(run())
         self.assertEqual(res, [i * 2 for i in range(1, 10)])
 
+    def test_async_gen_expression_incorrect(self):
+        err_msg = "'async for' requires an object with " \
+            "__aiter__ method, got int"
+        with self.assertRaisesRegex(TypeError, err_msg):
+            g = (x async for x in 42)
+
```

Comment:
I'd like a little more testing here. For example:

- What happens to objects that support `tp_iter`, but not the sequence protocol? What about vice versa?
- What about objects that support iteration synchronously, but not asynchronously? Vice versa?
- What if an object supports some other async operation (*i.e.*, a non-NULL `tp_as_async`), but not async iteration?

Reply:
I've added some tests.